net/graphsync_fetcher.go

```diff
@@ -0,0 +1,180 @@
+// using a Graphsync exchange to fetch tipsets recursively
+type GraphSyncFetcher struct {
+	exchange  GraphExchange
+	validator consensus.BlockSyntaxValidator
+	store     bstore.Blockstore
+	ssb       selector.SelectorSpecBuilder
+}
+
+// NewGraphSyncFetcher returns a GraphsyncFetcher wired up to the input Graphsync exchange and attached local
+// blockservice for reloading blocks in memory once they are returned
+func NewGraphSyncFetcher(ctx context.Context, exchange GraphExchange, blockstore bstore.Blockstore,
+	bv consensus.BlockSyntaxValidator) *GraphSyncFetcher {
+	gsf := &GraphSyncFetcher{
+		store:     blockstore,
+		validator: bv,
+		exchange:  exchange,
+		ssb:       selector.NewSelectorSpecBuilder(ipldfree.NodeBuilder()),
+	}
+	return gsf
+}
+
+// Graphsync can fetch a fixed number of tipsets from a remote peer recursively
+// with a single request. We don't know until we get all of the response whether
+// our final tipset was included in the response
+//
+// When fetching tipsets we try to balance performance for two competing cases:
+// - an initial chain sync that is likely to fetch lots and lots of tipsets
+// - a future update sync that is likely to fetch only a few
+//
+// To do this, the Graphsync fetcher starts fetching a single tipset at a time,
+// then gradually ramps up to fetch lots of tipsets at once, up to a fixed limit
+//
+// The constants below determine the maximum number of tipsets fetched at once
+// (maxRecursionDepth) and how fast the ramp up is (recursionMultipler)
+const maxRecursionDepth = 64
+const recursionMultiplier = 4
+
+// FetchTipSets gets Tipsets starting from the given tipset key and continuing till
+// the done function returns true or errors
+//
+// For now FetchTipSets operates in two parts:
+// 1. It fetches relevant blocks through Graphsync, which writes them to the block store
+// 2. It reads them from the block store and validates there syntax as blocks
+// and constructs a tipset
+// This does have a potentially unwanted side effect of writing blocks to the block store
+// that later don't validate (bitswap actually does this as well)
+// In the future, the blocks will be validated directly through graphsync as
+// go-filecoin migrates to the same IPLD library used by go-graphsync (go-ipld-prime)
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	err := gsf.fetchBlocks(ctx, cids, from)
+	if err != nil {
+		return nil, err
+	}
+	startingTipset, err := gsf.loadTipsetFromCids(ctx, cids)
+	if err != nil {
+		return nil, err
+	}
+	out := []types.TipSet{startingTipset}
+	isDone, err := done(startingTipset)
+	if err != nil {
+		return nil, err
+	}
+	recursionDepth := 1
+	ts := startingTipset
+	for !isDone {
+		err := gsf.fetchBlocksRecursively(ctx, ts.At(0).Cid(), from, recursionDepth)
```

Comment:
I think what is happening here is that we include the first block of our previously fetched tipset in our query in order to avoid the waste that comes from querying multiple starting cids.  If I've got that correct then this is great, but it is the kind of thing that looks like an error and will cause confusion and mistaken "I found a bug" moments for future readers.  It would be awesome to have a comment explaining that this is intentional and why we need it.